Rug approximations

Aras14HD opened the pull request:
I saw that you now integrated rug, and moved the approximation code over to its types, increasing the accuracy of the approximations. The precision is currently set to 1024 bits and can be controlled with the `math::FLOAT_PRECISION` constant.

This improvement means, that for all current approximations (current limits) only the last digit may be wrong. 

Todo:
- [x] Set new limits
- [x] Maybe change factorial::Factorial to not be limited to u128 inputs

Aras14HD commented:
Just checked for the new limits, they are beyond u128::MAX. One odd problem is that we will never get a digit approximation for single factorials. I might have to remove that test, or expand to bigger inputs (Integer).

Aras14HD commented:
I changed the input types, now there is no upper limit to the size of factorial people can calculate, for approximations the largest verifiable number was used. The new limit is 1000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000!
which is approximately 6.185167193060592 × 10^197565705518096748172348871081083394917705602994196333433885546216834135350791129225270775050661568251681293893255233696266358320712841036093430778935337187734147872913431329670406629130341173311668935

Here all digits are correct!

tolik518 commented:
Thank you very much so far, look awesome. I'll review it today in the evening (+8hrs from now)

tolik518 reviewed (changes requested):
@Aras14HD Excluding the small typo I only have one comment:
Do you thing we should use scientific notation for too long output? 
Since on very big number factorion won't manage to generate an answer under 10k characters.

See this example: https://www.reddit.com/r/test/comments/1izbtow/comment/mf4xar4/

Edit: I also see that the reply text in the example above is not optimal, lol

Aras14HD reviewed (commented):
Typo was fixed

Aras14HD commented:
I changed the truncation to apply to all factorial types and changed the behavior of oversize comments. Now the largest factorials are progressively removed until the comment is small enough.

Aras14HD commented:
While programming I noticed that the limit for subfactorials is relatively low. If it is due to performance, why not use the existing factorial function to calculate it as `!n = [n!/e]`. So it would just be:
```rust
(factorial(n,1) / Float::with_val(FLOAT_PRECISION, 1).exp()).to_integer()
```

tolik518 commented:
The limit is so low because of performance reasons iirc.

I have used malachite because when I implemented the function I was in a hurry and did no readings on how subfactorials work.

So it wasn't thought-out and every suggestion is welcome as always 😅

Aras14HD commented:
Tried it out, but it is not exact due to float precision, needs a 64 kB precision to even just pass the current tests. May still be something to think about if we want to enable approximate large subfactorials.

tolik518 commented:
Looks good to me. Thank you again for your contribution. The bot will get deployed after the tests ran after the merge.

Can't wait for the first reactions when people see it in the wild which numbers are possible now.
